extensions/src/main/java/org/pentaho/platform/plugin/services/metadata/PentahoAsyncThreadRunner.java

```diff
@@ -0,0 +1,153 @@
+/*!
+ *
+ * This program is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
+ * Foundation.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along with this
+ * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
+ * or from the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Lesser General Public License for more details.
+ *
+ *
+ * Copyright (c) 2022 Hitachi Vantara. All rights reserved.
+ *
+ */
+
+package org.pentaho.platform.plugin.services.metadata;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.pentaho.platform.api.engine.IPentahoSession;
+import org.pentaho.platform.engine.core.system.PentahoSessionHolder;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * Asynchronously executes a collection of tasks (ie {@link Callable} ). Contains logic that correctly
+ * sets the {@link IPentahoSession} for each spawned thread. Wrapper around {@link ExecutorService}.
+ */
+public class PentahoAsyncThreadRunner {
```

Comment:
This code seems to be generic and independent of _metadata_. Would it make sense to move it to a broader package? Or else, make it package private?

Reply:
Yes I can move it to another package. I did intend for other people to use it.

Reply:
from second PR review, will make this an inner class of SessionCachingMetadataDomainRepository.java

Reply:
see commit: https://github.com/pentaho/pentaho-platform/pull/5124/commits/4eb7eb5e6be8ae5e146ac8818babf2f3b5b60a20